src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/HttpVersionCompleter.cs

```diff
@@ -0,0 +1,39 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+#nullable enable
+
+using System.Collections;
+using System.Collections.Generic;
+using System.Diagnostics.CodeAnalysis;
+using System.Management.Automation;
+using System.Management.Automation.Language;
+
+namespace Microsoft.PowerShell.Commands
+{
+    /// <summary>
+    /// A completer for HTTP version names.
+    /// </summary>
+    [SuppressMessage("Microsoft.Performance", "CA1812:AvoidUninstantiatedInternalClasses", Justification = "Class is instantiated through late-bound reflection")]
+    internal sealed class HttpVersionCompleter : IArgumentCompleter
+    {
+        /// <inheritdoc/>
+        public IEnumerable<CompletionResult> CompleteArgument(string commandName, string parameterName, string wordToComplete, CommandAst commandAst, IDictionary fakeBoundParameters)
+        {
+            var valueToComplete = wordToComplete.Trim('\'');
+
+            var wordToCompletePattern = WildcardPattern.Get(string.IsNullOrWhiteSpace(valueToComplete) ? "*" : valueToComplete + "*", WildcardOptions.IgnoreCase);
+
+            foreach (var version in HttpVersionUtils.AllowedVersions)
+            {
+                if (wordToCompletePattern.IsMatch(version))
+                {
+                    var quotedVersion = Quoted(version);
+                    yield return new CompletionResult(quotedVersion, quotedVersion, CompletionResultType.ParameterValue, quotedVersion);
+                }
+            }
```

Comment:
I think it is more readable to show version names to users, not values.
So please replace `var` with `string` and remove Quoted() since no spaces are in the names.

Reply:
1.0 treated as double without quoting
Do you still thinking that we must instantiate Version in all cases during transformation? I wouldn't to change current behavior for Set-StrictMode

Reply:
> 1.0 treated as double without quoting

If we want to use string names (like HTTP3) it is not problem. Direct using double is also not problem.

> Do you still thinking that we must instantiate Version in all cases during transformation? I wouldn't to change current behavior for Set-StrictMode

Current implementation has implicit behaviors in transformation attribute. I believe we could do the work explicitly and this will allow us to use dry principle for new (HttpVersion) attribute.

Reply:
Ilya, I'm confused )
Are you ok with completion 1.0, 1.1, 2.0, 3.0? Or you want to change it to HTTP1, HTTP1.1, etc?
Are you ok with my current implementation of ArgumentToVersionTranformationAttribute? I made class extensible

Reply:
We usually discuss this in issue before PR. But since it is still draft we can consider alternatives.
I think `-HttpVersion HTTP21` looks more readable than `-HttpVersion 2.1`. This follows "Latest" in Set-Strict. If you prefer second one (double) I have no objections since we can add first one later without a breaking change. Also we can refactor existing code later (Ideally we could have one universal parametrized attribute for both cases.)

So you have two options to choose from:
- simply add enhancement as you started
- continue to implement common solution with code refactoring.

Reply:
As I understood, you propose to do something like this (to be universal enough):

```csharp
public TransformationAttribute
{

    private readonly IReadOnlyDictionary<string, Version> _map;

    public TransformationAttribute(Type mappingProvider)
    {
        _map = GetValues(mappingProvider);
    }

    // other code
}

// Invoke-WebRequest
[TransformationAttribute(typeof(HttpVersionMapping))]
// Set-StrictMode
[TransformationAttribute(typeof(SetStrictModeVersionMapping))]
```

According to [specification](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/language-specification/attributes#attribute-parameter-types) we cannot create universal parametrized implementation other than via `Type` argument with implicit convention.

And I think this not worth to do it. At least for now.

Reply:
I hope this is it )